peerconnection.go

```diff
@@ -1878,72 +1675,36 @@ func (pc *PeerConnection) startTransports(iceRole ICERole, dtlsRole DTLSRole, re
 		return
 	}
 
-	pc.openSRTP()
+	pc.startRTPReceivers(trackDetailsFromSDP(pc.log, pc.RemoteDescription().parsed), currentTransceivers)
+	pc.startRTPSenders(currentTransceivers)
+	pc.drainSRTP()
+	pc.startSCTP()
+}
 
-	for _, tranceiver := range pc.GetTransceivers() {
-		if tranceiver.Sender != nil {
-			err = tranceiver.Sender.Send(RTPSendParameters{
-				Encodings: RTPEncodingParameters{
-					RTPCodingParameters{
-						SSRC:        tranceiver.Sender.track.SSRC(),
-						PayloadType: tranceiver.Sender.track.PayloadType(),
-					},
-				}})
-			if err != nil {
-				pc.log.Warnf("Failed to start Sender: %s", err)
-			}
+func (pc *PeerConnection) startRenegotation(currentTransceivers []*RTPTransceiver) {
+	trackDetails := trackDetailsFromSDP(pc.log, pc.RemoteDescription().parsed)
+	for _, t := range currentTransceivers {
+		if t.Receiver() == nil || t.Receiver().Track() == nil {
+			continue
+		} else if _, ok := trackDetails[t.Receiver().Track().ssrc]; ok {
+			continue
 		}
-	}
 
-	go pc.drainSRTP()
-
-	// Start sctp
-	if err = pc.sctpTransport.Start(SCTPCapabilities{
-		MaxMessageSize: 0,
-	}); err != nil {
-		pc.log.Warnf("Failed to start SCTP: %s", err)
-		if err = pc.sctpTransport.Stop(); err != nil {
-			pc.log.Warnf("Failed to stop SCTPTransport: %s", err)
+		if err := t.Receiver().Stop(); err != nil {
+			pc.log.Warnf("Failed to stop RtpReceiver: %s", err)
+			continue
 		}
 
-		return
-	}
-
-	// DataChannels that need to be opened now that SCTP is available
-	// make a copy we may have incoming DataChannels mutating this while we open
-	pc.sctpTransport.lock.RLock()
-	dataChannels := append([]*DataChannel{}, pc.sctpTransport.dataChannels...)
-	pc.sctpTransport.lock.RUnlock()
-
-	var openedDCCount uint32
-	for _, d := range dataChannels {
-		if d.ReadyState() == DataChannelStateConnecting {
-			err := d.open(pc.sctpTransport)
-			if err != nil {
-				pc.log.Warnf("failed to open data channel: %s", err)
-				continue
-			}
-			openedDCCount++
+		receiver, err := pc.api.NewRTPReceiver(t.Receiver().kind, pc.dtlsTransport)
```

Comment:
In my understanding, this is called if a transceiver `t` in the current list of transceivers is removed from remote description.
I'm not sure why new receiver is set to the transceiver which is removed.
I think `t` should be deleted from `pc.rtpTransceivers` like:
```go
func (pc *PeerConnection) startRenegotation(currentTransceivers []*RTPTransceiver) {
  trackDetails := trackDetailsFromSDP(pc.log, pc.RemoteDescription().parsed)
  var newTransceivers []*RTPTransceiver
  pc.mu.Lock()
  for _, t := range currentTransceivers {
    if t.Receiver() == nil || t.Receiver().Track() == nil {
      if t.Sender() != nil {
        newTransceivers = append(newTransceivers, t)
      } else {
        t.setDirection(RTPTransceiverDirectionInactive)
      }
      continue
    } else if _, ok := trackDetails[t.Receiver().Track().ssrc]; ok {
      newTransceivers = append(newTransceivers, t)
      continue
    }
    
    if err := t.Receiver().Stop(); err != nil {
      pc.log.Warnf("Failed to stop RtpReceiver: %s", err)
      continue
    }
    t.setReceiver(nil)
    if t.Sender() != nil {
      newTransceivers = append(newTransceivers, t)
    } else {
      t.setDirection(RTPTransceiverDirectionInactive)
    } 
  }
  pc.rtpTransceivers = newTransceivers
  pc.mu.Unlock()
  
  pc.startRTPSenders(currentTransceivers)
  pc.startRTPReceivers(trackDetails, currentTransceivers)
}
```

Reply:
(Let me preface this with I am still learning the Transceiver API, so this might be wrong)

* trackDetailsFromSDP just looks up the incoming SSRCes. So a Transceiver could just have had its SSRC flipped (remote did a TrackRemove -> TrackAdd). It isn't really transceiver aware. 
* Transceivers can never deleted/removed. They only have their directions changed.

We should probably put this code in a dedicated function like 'pc.stopRTPReceivers`, its purpose is just to loop through and stop incoming stuff.

Later we will stop our `Senders` if we are answering. We loop over the remotes transceivers and if they provide none that satisfy the local we don't tell the remote about one of our senders.

Reply:
Thanks, I finally got it. My understanding around Track and Transceiver had been confused.